File: ios/Plugin/GlassfyPlugin.swift
import Capacitor
import Foundation
import Glassfy
import GlassfyGlue

@objc(GlassfyPlugin)
public class GlassfyPlugin: CAPPlugin {
    private weak var paywallViewController: UIViewController?
    private var paywallListener: PaywallListener?
    
    func convertResponseFromGlassfyGlue(_ call: CAPPluginCall) -> ([String: Any]?, Error?) -> Void {
        return { value, error in
            if let error {
                call.reject(error.localizedDescription)
                return
            }
            if let value {
                call.resolve(value)
            } else {
                call.resolve()
            }
        }
    }    
    
    @objc func sdkVersion(_ call: CAPPluginCall) {
        GlassfyGlue.sdkVersion (completion: self.convertResponseFromGlassfyGlue(call));
    }

    @objc func initialize(_ call: CAPPluginCall) {
        guard let apiKey = call.getString("apiKey"),
              let watcherMode = call.getBool("watcherMode")
        else {
            call.reject("invalid initialize parameters")
            return
        }
        GlassfyGlue.initialize(
            withApiKey: apiKey, 
            watcherMode: watcherMode, 
            crossPlatformSdkFramework: "capacitor",
            crossPlatformSdkVersion: "3.0.0",
            withCompletion: self.convertResponseFromGlassfyGlue(call)
        )
    }
    
    @objc func setLogLevel(_ call: CAPPluginCall) {
        guard let logLevel = call.getInt("logLevel") else {
            call.reject("invalid login parameters")
            return
        }
        if let ll = Glassfy.LogLevel(rawValue: UInt(logLevel)) {
            Glassfy.log(level: ll)
        }
    }

    @objc func offerings(_ call: CAPPluginCall) {
        GlassfyGlue.offerings(completion: self.convertResponseFromGlassfyGlue(call))
    }

    @objc func purchaseHistory(_ call: CAPPluginCall) {
        GlassfyGlue.purchaseHistory(completion: self.convertResponseFromGlassfyGlue(call))
    }
    
    @objc func permissions(_ call: CAPPluginCall) {
        GlassfyGlue.permissions(completion: self.convertResponseFromGlassfyGlue(call));
    }
    
    @objc func skuWithId(_ call: CAPPluginCall) {
        guard let identifier = call.getString("identifier") else {
            call.reject("invalid skuWithIdentifier parameters")
            return
        }
        GlassfyGlue.sku(withId: identifier, withCompletion: self.convertResponseFromGlassfyGlue(call));
    }
    
    @objc func skuWithIdAndStore(_ call: CAPPluginCall) {
        guard let identifier = call.getString("identifier") else {
            call.reject("invalid skuWithIdAndStore 'identifier' parameters")
            return
        }
        guard let storecode = call.getInt("identifier") else {
            call.reject("invalid skuWithIdAndStore 'store' parameters")
            return
        }
        let store:Glassfy.Store
        switch storecode {
        case 1: store = Glassfy.Store.appStore
        case 2: store = Glassfy.Store.playStore
        case 3: store = Glassfy.Store.paddle
        case 4: store = Glassfy.Store.stripe
        case 5: store = Glassfy.Store.glassfy
        default:
            call.reject("invalid skuWithIdAndStore 'store' parameters")
            return
        }        
        GlassfyGlue.sku(withId: identifier, store: store, completion: self.convertResponseFromGlassfyGlue(call));
    }
    
    @objc func connectCustomSubscriber(_ call: CAPPluginCall) {
        guard let subscriberId = call.getString("subscriberId") else {
            call.reject("invalid subscriberId parameters")
            return
        }
        GlassfyGlue.connectCustomSubscriber(subscriberId, completion: self.convertResponseFromGlassfyGlue(call));
    }

    @objc func connectPaddleLicenseKey(_ call: CAPPluginCall) {
        guard let licenseKey = call.getString("licenseKey") else {
            call.reject("invalid licenseKey parameter")
            return
        }
        guard let force = call.getBool("force") else {
            call.reject("invalid force parameter")
            return
        }
        GlassfyGlue.connectPaddleLicenseKey(licenseKey, force: force , completion: self.convertResponseFromGlassfyGlue(call));
    }

    @objc func connectGlassfyUniversalCode(_ call: CAPPluginCall) {
        guard let universalCode = call.getString("universalCode") else {
            call.reject("invalid universalCode parameter")
            return
        }
        guard let force = call.getBool("force") else {
            call.reject("invalid force parameter")
            return
        }
        GlassfyGlue.connectGlassfyUniversalCode(universalCode, force: force , completion: self.convertResponseFromGlassfyGlue(call));
    }
    
    @objc func setEmailUserProperty(_ call: CAPPluginCall) {
        guard let email = call.getString("email") else {
            call.reject("invalid email parameters")
            return
        }
        GlassfyGlue.setEmailUserProperty(email, withCompletion: self.convertResponseFromGlassfyGlue(call));
    }
    
    @objc func setDeviceToken(_ call: CAPPluginCall) {
        guard let token = call.getString("token") else {
            call.reject("invalid setDeviceToken parameters")
            return
        }
        GlassfyGlue.setDeviceToken(token, withCompletion: self.convertResponseFromGlassfyGlue(call));
    }

    @objc func setExtraUserProperty(_ call: CAPPluginCall) {
        guard let extra = call.getObject("extra") else {
            call.reject("invalid setDeviceToken parameters")
            return
        }
        GlassfyGlue.setExtraUserProperty(extra, withCompletion: self.convertResponseFromGlassfyGlue(call));
    }

    @objc func getUserProperty(_ call: CAPPluginCall) {        
        GlassfyGlue.getExtraUserProperty(completion: self.convertResponseFromGlassfyGlue(call));        
    }

    @objc func purchaseSku(_ call: CAPPluginCall) {
        guard let sku = call.getObject("sku") else {
            call.reject("Invalid SKU")
            return
        }                
        GlassfyGlue.purchaseSku(sku, withCompletion: self.convertResponseFromGlassfyGlue(call));
    }
    
    @objc func restorePurchases(_ call: CAPPluginCall) {
        GlassfyGlue.restorePurchases(completion: self.convertResponseFromGlassfyGlue(call));
    }

    @objc func setAttribution(_ call: CAPPluginCall) {
        guard let type = call.getInt("type") else {
            call.reject("invalid/missing Type")
            return
        }
        guard let value = call.getString("value") else {
            call.reject("invalid/missing value")
            return
        }
        GlassfyGlue.setAttributionType(NSNumber(integerLiteral: type), value:value, completion: self.convertResponseFromGlassfyGlue(call));
    }
    
    @objc func setAttributions(_ call: CAPPluginCall) {
        guard let items = call.getArray("items") else {
            call.reject("invalid/missing items")
            return
        }
        GlassfyGlue.setAttributions(items, completion: self.convertResponseFromGlassfyGlue(call));
    }
    
    @objc func _paywall(_ call: CAPPluginCall) {
        guard let remoteConfig = call.getString("remoteConfig") else {
            call.reject("Invalid/missing remoteConfig")
            return
        }
        guard paywallViewController == nil else {
            call.reject("Can only show one paywall at a time, please call `GlassfyPaywall.close()`")
            return
        }
        Task {
            await showPaywall(id: remoteConfig, then: call)
        }
    }
    
    private func showPaywall(id remoteConfig: String, then call: CAPPluginCall) async {
        do {
            let paywall = try await Glassfy.paywall(id: remoteConfig)
            let viewController = try await paywall.loadViewController()
            let listener = CapacitorPaywallListener { self.notifyListeners($0, data: $1) }
            
            self.paywallViewController = viewController
            self.paywallListener = listener
            
            Task { @MainActor in
                viewController.install(listener)
                bridge?.viewController?.present(viewController, animated: true)                
                call.resolve(["result": "sucess"])
            }
        } catch {
            call.reject("Failed to show paywall, \(error)")
        }
    }
    
    @objc func _closePaywall(_ call: CAPPluginCall) {
        Task { @MainActor in
            paywallViewController?.dismiss(animated: true)
            paywallViewController = nil
            paywallListener = nil
            call.resolve(["result": "sucess"])
        }
    }
    
    @objc func _openUrl(_ call: CAPPluginCall) {
        guard let urlString = call.getString("url"), let url = URL(string: urlString) else {
            call.reject("Invalid/missing url")
            return
        }
        UIApplication.shared.open(url)
        call.resolve(["result": "sucess"])
    }
}

File: android/src/main/java/io/glassfy/capacitor/plugin/GlassfyPlugin.kt
package io.glassfy.capacitor.plugin

import android.content.Intent
import android.net.Uri
import androidx.fragment.app.DialogFragment
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.PluginMethod
import com.getcapacitor.PluginCall
import io.glassfy.glue.GlassfyGlue
import io.glassfy.paywall.GlassfyPaywall
import kotlinx.coroutines.MainScope
import org.json.JSONArray
import org.json.JSONObject

@CapacitorPlugin(name = "Glassfy")
class GlassfyPlugin : Plugin() {
    private var paywallListener: CapacitorPaywallListener? = null
    private var paywallFragment: DialogFragment? = null

    fun JSONObject.toMap(): Map<String, *> = keys().asSequence().associateWith {
        when (val value = this[it]) {
            is JSONArray -> {
                val map = (0 until value.length()).associate { Pair(it.toString(), value[it]) }
                JSONObject(map).toMap().values.toList()
            }
            is JSONObject -> value.toMap()
            JSONObject.NULL -> null
            else -> value
        }
    }

    private fun pluginCompletion(call: PluginCall, value: String?, error: String?) {
        if (error != null) {
            call.reject(error)
            return
        }
        if (value==null) {
            call.resolve()
            return;
        }
        val jo = value.let { JSObject(it) }
        call.resolve(jo)
    }

    @PluginMethod
    fun sdkVersion(call: PluginCall) {
        GlassfyGlue.sdkVersion { value, error -> pluginCompletion(call, value, error) }
    }

    @PluginMethod
    fun initialize(call: PluginCall) {
        var apiKey = call.getString("apiKey")
        var watcherMode = call.getBoolean("watcherMode")
        if (apiKey == null || watcherMode == null) {
            call.reject("invalid parameter")
            return
        }
        GlassfyGlue.initialize(
            this.getContext(),
            apiKey,
            watcherMode,
            "capacitor",
            "3.3.3"
        ) { value, error -> pluginCompletion(call, value, error) }
    }

    @PluginMethod
    fun setLogLevel(call: PluginCall) {
        var logLevel = call.getInt("logLevel")

        if (logLevel == null ) {
            call.reject("invalid parameter")
            return
        }
        GlassfyGlue.setLogLevel(logLevel);
        call.resolve()
    }

    @PluginMethod
    fun offerings(call: PluginCall) {
        GlassfyGlue.offerings { value, error -> pluginCompletion(call, value, error) }
    }

    @PluginMethod
    fun purchaseHistory(call: PluginCall) {
        GlassfyGlue.purchaseHistory { value, error -> pluginCompletion(call, value, error) }
    }

    @PluginMethod
    fun permissions(call: PluginCall) {
        GlassfyGlue.permissions { value, error -> pluginCompletion(call, value, error) }
    }

    @PluginMethod
    fun skuWithId(call: PluginCall) {
        var identifier = call.getString("identifier")
        if (identifier == null ) {
            call.reject("invalid parameter")
            return
        }
        GlassfyGlue.skuWithId(identifier) { value, error -> pluginCompletion(call, value, error) }
    }

    @PluginMethod
    fun skuWithIdAndStore(call: PluginCall) {
        var identifier = call.getString("identifier")
        var store = call.getInt("store")

        if (identifier == null || store == null ) {
            call.reject("invalid parameter")
            return
        }
        GlassfyGlue.skuWithIdAndStore(identifier, store) { value, error ->
            pluginCompletion(
                call,
                value,
                error
            )
        }
    }

    @PluginMethod
    fun connectCustomSubscriber(call: PluginCall) {
        var subscriberId = call.getString("subscriberId")
        if (subscriberId == null ) {
            call.reject("invalid parameter")
            return
        }
        GlassfyGlue.connectCustomSubscriber(subscriberId) { value, error ->
            pluginCompletion(
                call,
                value,
                error
            )
        }
    }

    @PluginMethod
    fun connectPaddleLicenseKey(call: PluginCall) {
        var licenseKey = call.getString("licenseKey")
        var force = call.getBoolean("force")

        if (licenseKey == null || force == null ) {
            call.reject("invalid parameter")
            return
        }
        GlassfyGlue.connectPaddleLicenseKey(licenseKey, force) { value, error ->
            pluginCompletion(
                call,
                value,
                error
            )
        }
    }

    @PluginMethod
    fun connectGlassfyUniversalCode(call: PluginCall) {
        var universalCode = call.getString("universalCode")
        var force = call.getBoolean("force")

        if (universalCode == null || force == null ) {
            call.reject("invalid universalCode parameter")
            return
        }
        GlassfyGlue.connectGlassfyUniversalCode(universalCode, force) { value, error ->
            pluginCompletion(
                call,
                value,
                error
            )
        }
    }

    @PluginMethod
    fun setEmailUserProperty(call: PluginCall) {
        var email = call.getString("email")
        if (email == null ) {
            call.reject("invalid parameter")
            return
        }
        GlassfyGlue.setEmailUserProperty(email) { value, error ->
            pluginCompletion(
                call,
                value,
                error
            )
        }
    }

    @PluginMethod
    fun setExtraUserProperty( call: PluginCall) {
        var extra = call.getObject("extra").toMap()

        val map = HashMap<String,String>()
        var validMap = true
        extra.forEach { (key, value) ->
            if (value is String) {
                map[key] = value
            } else {
                validMap = false;
            }
        }
        if (validMap == false) {
            call.reject("invalid parameter")
            return
        }

        GlassfyGlue.setExtraUserProperty(map) { value, error ->
            pluginCompletion(
                call,
                value,
                error
            )
        }
    }

    @PluginMethod
    fun getUserProperties(call: PluginCall) {
        GlassfyGlue.getExtraUserProperty() { value, error -> pluginCompletion(call, value, error) }
    }

    @PluginMethod
    fun purchaseSku(call: PluginCall) {
        if (activity == null) {
            call.reject("Invalid Android Activity", "Invalid Android Activity")
            return
        }

        val sku = call.getObject("sku", null)
        if (sku == null) {
            call.reject("missing sku parameter")
            return
        }
        val skuId = sku.getString("skuId")
        if (skuId == null) {
            call.reject("invalid sku object")
            return
        }

        val skuToUpgrade = call.getObject("skuToUpgrade", null)

        var subscriptionUpdateId: String? = null
        var replacementMode: Int? = null
        if (skuToUpgrade != null) {
            subscriptionUpdateId = skuToUpgrade.getString("skuId")
            if (subscriptionUpdateId == null) {
                call.reject("Invalid skuToUpgrade")
                return
            }
            replacementMode = call.getInt("replacementMode")
            if (replacementMode == null) {
                replacementMode = 1
            }
        }
        GlassfyGlue.purchaseSku(
            activity,
            skuId,
            subscriptionUpdateId,
            replacementMode
        ) { v, e -> pluginCompletion(call, v, e) }
    }

    @PluginMethod
    fun restorePurchases(call: PluginCall) {
        GlassfyGlue.restorePurchases() { value, error -> pluginCompletion(call, value, error) }
    }

    @PluginMethod
    fun storeInfo(call: PluginCall) {
        GlassfyGlue.storeInfo() { value, error -> pluginCompletion(call, value, error) }
    }

    @PluginMethod
    fun setDeviceToken(call: PluginCall) {
        call.resolve()
    }

    @PluginMethod
    fun setAttribution(call: PluginCall) {
        val type = call.getInt("type")
        if (type == null ) {
            call.reject("invalid/missing parameter 'type'")
            return
        }
        val attribValue = call.getString("value")
        if (attribValue == null ) {
            call.reject("invalid/missing parameter 'value'")
            return
        }

        GlassfyGlue.setAttribution(type,attribValue) { value, error ->
            pluginCompletion(
                call,
                value,
                error
            )
        }
    }

    @PluginMethod
    fun setAttributions(call: PluginCall) {
        val items = call.getArray("items")
        if (items == null) {
            call.reject("invalid/missing parameter 'items'")
            return
        }

        val listItems =  mutableListOf<Map<String,Any?>>()
        for (i in 0 until items.length()) {
            val item = items.getJSONObject(i).toMap()

            listItems.add(item)
        }

        GlassfyGlue.setAttributions(listItems) { value, error ->
            pluginCompletion(
                call,
                value,
                error
            )
        }
    }

    @PluginMethod
    fun _paywall(call: PluginCall) {
        if (paywallFragment != null) {
            call.reject("Only one paywall can be shown at a time, please call `GlassfyPaywall.close()`")
            return
        }
        val awaitLoading = call.getBoolean("awaitLoading") ?: false
        val remoteConfig = call.getString("remoteConfig")
        if (remoteConfig == null) {
            call.reject("invalid/missing parameter 'remoteConfig'")
            return
        }

        val listener = CapacitorPaywallListener(this)
        paywallListener = listener
        GlassfyPaywall.fragment(this.getContext(), remoteConfig, awaitLoading) { paywall, error ->
            MainScope().run {
                paywall?.setCloseHandler(listener.onClose)
                paywall?.setPurchaseHandler(listener.onPurchase)
                paywall?.setRestoreHandler(listener.onRestore)
                paywall?.setLinkHandler(listener.onLink)
                paywall?.show(bridge.activity.supportFragmentManager, "paywall")
            }
            paywallFragment = paywall
            pluginCompletion(call, null, null)
        }
    }

    @PluginMethod
    fun _openUrl(call: PluginCall) {
        val urlString = call.getString("url")
        if (urlString == null) {
            call.reject("invalid/missing parameter 'url'")
            return
        }
        try {
            val i = Intent(Intent.ACTION_VIEW).apply { data = Uri.parse(urlString) }
            bridge.activity.startActivity(i)
        } catch(_: Exception) {
            call.reject("invalid parameter 'url'")
        }
    }

    @PluginMethod
    fun _closePaywall(call: PluginCall) {
        MainScope().run {
            paywallFragment?.dismiss()
            paywallFragment = null
            paywallListener = null
        }
        pluginCompletion(call, null, null)
    }

    internal fun sendEvent(eventName: String, payload: JSONObject) {
        val jsObject = JSObject(payload.toString());
        notifyListeners(eventName, jsObject);
    }
}